src/swap/defi/thorchain.js

```diff
@@ -0,0 +1,611 @@
+        new Date(Date.now() + EXPIRATION_MS),
+        tx.txid
+      )
+    }
+  }
+  return out
+}
+
+type BuildSwapMemoParams = {
+  chain: string,
+  asset: string,
+  address: string,
+  limit: string,
+  affiliateAddress: string,
+  points: string
+}
+
+const calcSwapFrom = async (
+  params: {
+    fromWallet: EdgeCurrencyWallet,
+    fromCurrencyCode: string,
+    toWallet: EdgeCurrencyWallet,
+    toCurrencyCode: string,
+    nativeAmount: string,
+    sourcePool: Pool,
+    destPool: Pool,
+    volatilitySpreadFinal: string,
+    affiliateFee: string,
+    feeInDestCurrency: string,
+    dontCheckLimits?: boolean
+  },
+  clog: Function
+): Promise<{
+  fromNativeAmount: string,
+  fromExchangeAmount: string,
+  toNativeAmount: string,
+  toExchangeAmount: string,
+  limit: string
+}> => {
+  const {
+    fromWallet,
+    fromCurrencyCode,
+    toWallet,
+    toCurrencyCode,
+    nativeAmount,
+    sourcePool,
+    destPool,
+    volatilitySpreadFinal,
+    affiliateFee,
+    feeInDestCurrency
+  } = params
+  const fromNativeAmount = nativeAmount
+
+  // Get exchange rate from source to destination asset
+  const fromExchangeAmount = await fromWallet.nativeToDenomination(
+    nativeAmount,
+    fromCurrencyCode
+  )
+
+  clog(`fromExchangeAmount: ${fromExchangeAmount}`)
+
+  const result = calcDoubleSwapOutput(
+    Number(mul(fromExchangeAmount, THOR_LIMIT_UNITS)),
+    sourcePool,
+    destPool
+  )
+
+  let toExchangeAmount = div(
+    result.toString(),
+    THOR_LIMIT_UNITS,
+    DIVIDE_PRECISION
+  )
+  clog(`toExchangeAmount: ${toExchangeAmount}`)
+
+  const subVolatility = mul(toExchangeAmount, volatilitySpreadFinal)
+  const subAffiliateFee = mul(toExchangeAmount, affiliateFee)
+
+  toExchangeAmount = sub(toExchangeAmount, subVolatility)
+  clog(
+    `toExchangeAmount w/volatilitySpread of ${mul(
+      volatilitySpreadFinal,
+      '100'
+    )}%: ${toExchangeAmount}`
+  )
+  toExchangeAmount = sub(toExchangeAmount, subAffiliateFee)
+  clog(
+    `toExchangeAmount w/affiliate fee of ${mul(
+      affiliateFee,
+      '100'
+    )}%: ${toExchangeAmount}`
+  )
+  toExchangeAmount = sub(toExchangeAmount, feeInDestCurrency)
+  clog(
+    `toExchangeAmount w/network fee of ${feeInDestCurrency}: ${toExchangeAmount}`
+  )
+
+  const toNativeAmount = await toWallet.denominationToNative(
+    toExchangeAmount,
+    toCurrencyCode
+  )
+  clog(`toNativeAmount: ${toNativeAmount}`)
+  const limit = toFixed(mul(toExchangeAmount, THOR_LIMIT_UNITS), 0, 0)
+  clog(`limit: ${limit}`)
+
+  return {
+    fromNativeAmount,
+    fromExchangeAmount,
+    toNativeAmount,
+    toExchangeAmount,
+    limit
+  }
+}
+
+const buildSwapMemo = (params: BuildSwapMemoParams): string => {
+  const { chain, asset, address, limit, affiliateAddress, points } = params
+  // affiliate address could be a thorname, and the minimum received is not set in this example.
+  return `=:${chain}.${asset}:${address}:${limit}:${affiliateAddress}:${points}`
```

Comment:
This is an ASCII string, but we agreed to pass binary-data memos in hex. Where is the `base16.stringify(utf8.parse(...))`? I assume Thorchain wants the OP_RETURN or EVM extra data in UTF-8, which means we should make that conversion in the Thorchain plugin, not in the currency plugins.

Maybe we should have made the memo parameter an `Uint8Array`, just to be extra-explicit about these data-encoding things.

Reply:
For the OP_RETURN we just pass an ascii string. That's the API for OP_RETURN as implemented. For EVM data, I convert the string to base16 here

https://github.com/EdgeApp/edge-exchange-plugins/pull/209/commits/6c74e38114a34392c2744befd3480ab91bfa0696#diff-d025ca361f3e8068e0bfdd23f1c3dcca99c3d2e5507cfa49c7503857d8e51f1eR434